activerecord/lib/active_record/middleware/strict_safe_http_methods.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+# :markup: markdown
+
+module ActiveRecord
+  module Middleware
+    # # Strict Safe HTTP Methods \Middleware
+    #
+    # This middleware prevents database writes while serving a "safe" HTTP request (`GET`, `HEAD`, or
+    # `OPTIONS`) by raising an `ActiveRecord::ReadOnlyError` exception if a database write is
+    # attempted.
+    #
+    # Disabled by default, this middleware can be added by setting
+    # `config.active_record.strict_safe_http_methods` to `true`.
+    class StrictSafeHTTPMethods
+      def initialize(app)
+        @app = app
+      end
+
+      def call(env)
+        request = ActionDispatch::Request.new(env)
+
+        if request.get? || request.head? || request.options?
```

Comment:
The [spec](https://httpwg.org/specs/rfc9110.html#safe.methods) also mentions `TRACE`. Should it be included or should the set of safe methods be configurable?

Reply:
I don't think the routing layer supports `TRACE`, for example

https://github.com/flavorjones/rails/blob/91b0e30e146d0af8d12c8d23053167efa7ef45c7/actionpack/lib/action_dispatch/routing.rb#L260 only specifies

```
    HTTP_METHODS = [:get, :head, :post, :patch, :put, :delete, :options] # :nodoc:
```